### PR TITLE
bpo-31061: fix crash in asyncio speedup module

### PR DESCRIPTION
**Pull request by thehesiod**
https://bugs.python.org/issue31061

follows what was done in: https://bugs.python.org/issue26617
fallout in: http://bugs.python.org/issue31095

### CODE REVIEW

**Comment by thehesiod** [comment]
this is my first cpython PR, let me know what else needs to be done to get this into any other releases it's required in.

**Comment by thehesiod** [comment]
not sure why the build failed, doesn't seem related, however it is in the same file

File: Modules/_asynciomodule.c
Comment (methane) [inline Modules/_asynciomodule.c]: Is this really needed?

Comment (methane) [reply Modules/_asynciomodule.c]: sorry, you're right. it's needed.

Comment (thehesiod) [reply Modules/_asynciomodule.c]: I'm just following what was done in https://bugs.python.org/issue26617. I think it's worth someone grepping through the whole codebase, I saw several other places that maybe needed this.  It seems like every dealloc should have a PyObject_GC_UnTrack at the top?

**Comment by methane** [comment]
Since `_asinciomodule` is compiled with python interpreter, it can use private macro APIs
(`_PyObject_GC_TRACK` and `_PyObject_GC_UNTRACK`).

**Comment by methane** [comment]
Oh, sorry, macro version API broke Windows build and you used public APIs to fix it.

**Review by methane — APPROVED** [review]

File: Modules/_asynciomodule.c
Comment (methane) [reply Modules/_asynciomodule.c]: Maybe, noddy4 example should do it at first :)

Comment (thehesiod) [reply Modules/_asynciomodule.c]: btw how do we get this into 3.6.3 as well?

Comment (methane) [reply Modules/_asynciomodule.c]: > It seems like every dealloc should have a PyObject_GC_UnTrack at the top?

When the type has `Py_TPFLAGS_HAVE_GC`.
`lru_cache_dealloc` in `_functoolsmodule` seems need it.

Comment (methane) [reply Modules/_asynciomodule.c]: I filed an issue for that. http://bugs.python.org/issue31095

Comment (thehesiod) [reply Modules/_asynciomodule.c]: k, should I remove the LRU change I added here?

File: Modules/_asynciomodule.c
Comment (methane) [inline Modules/_asynciomodule.c]: Please move PyObject_GC_UnTrack after this if block, and remove Track and Untrack in the block.

**Review by methane — CHANGES_REQUESTED** [review]

File: Modules/_asynciomodule.c
Comment (methane) [reply Modules/_asynciomodule.c]: Yes, please.  asyncio speedup module is introduced from Python 3.6.
It make easy to backport.

File: Modules/_asynciomodule.c
Comment (thehesiod) [reply Modules/_asynciomodule.c]: but that would mean this (https://github.com/python/cpython/commit/34fae03cd6c9e304e02c571b3bf9e8df0cfe76be)  is wrong too then no? 

because it first does an UnTrack: https://github.com/python/cpython/commit/34fae03cd6c9e304e02c571b3bf9e8df0cfe76be#diff-c3cf251f16d5a03a9e7d4639f2d6f998L1113

With the same Track/Untrack pattern here:
https://github.com/python/cpython/commit/34fae03cd6c9e304e02c571b3bf9e8df0cfe76be#diff-c3cf251f16d5a03a9e7d4639f2d6f998R1129

**Comment by methane** [comment]
Could you add news entry?
https://devguide.python.org/committing/#what-s-new-and-news-entries

**Comment by thehesiod** [comment]
added news and reverted LRU change

File: Modules/_asynciomodule.c
Comment (methane) [reply Modules/_asynciomodule.c]: current code:
```
UnTrack()
Track()
PyObject_CallFinalizerFromDealloc()
UnTrack()
```

I proposed:

```
PyObject_CallFinalizerFromDealloc()
UnTrack()
```

34fae03 does other things between UnTrack() and Track()
But this code does only `Future_CheckExact()` and it is safe.

Comment (thehesiod) [reply Modules/_asynciomodule.c]: k, I'm claiming ignorance since I don't know what these macros achieve :)  I'd personally feel safer if we had a test case, do you know how to create a testcase to cause this to crash?  Mine takes a day to reproduce in a production environment :(

Comment (methane) [reply Modules/_asynciomodule.c]: CPython's circular reference GC uses doubly linked list to track object which can be member of circular reference.
`PyObject_GC_UnTrack()` removes object from the list, and `PyObject_GC_Track()` insert the object to the list.

Despite very undetarministic multithreading, I can cause SEGV quickly with this code:

```
import asyncio
import gc
gc.set_debug(gc.DEBUG_STATS)

class Evil:
    def __del__(self):
        gc.collect()

def main():
    f = asyncio.Future()
    f.set_result(Evil())

for i in range(100):
    main()
```

Comment (thehesiod) [reply Modules/_asynciomodule.c]: nice!  I'll add testcases and add your suggestions, thanks so much for all the help!

Comment (thehesiod) [reply Modules/_asynciomodule.c]: k, added FutureObj testcase, have a good test case for the TaskObj?  Not quite sure how to trigger it.

**Comment by thehesiod** [comment]
should we put as many fixes as we can in this PR, or create another PR with as many as we can?  I need the defaultdict one too...and maybe even others as who knows what third party libraries do.  I feel like we've opened a pandoras box of crashers :)

**Comment by methane** [comment]
I already created #2974.
More pull request for updating document is welcome.

File: Include/objimpl.h
Comment (methane) [inline Include/objimpl.h]: Please remove it.
We can reach the document easily with searching API name.

**Comment by 1st1** [comment]
The latest revision LGTM. But can you use `blurb` to generate the NEWS entry?

**Comment by thehesiod** [comment]
@1st1 should I remove the one in NEWS?  Also was hoping I could add a testcase for Task but couldn't figure out how to repro it, especially given Task inherits from Future so I don't want to trigger the Future's crash

File: Misc/NEWS
Comment (1st1) [inline Misc/NEWS]: Last nit: the change in Misc/NEWS needs to be reverted, we need only blurb file.

**Comment by 1st1** [comment]
> @1st1 should I remove the one in NEWS? 

Yes, please!

> Also was hoping I could add a testcase for Task but couldn't figure out how to repro it, especially given Task inherits from Future so I don't want to trigger the Future's crash

I'm trying to come up with a test.

**Comment by 1st1** [comment]
The following test crashes unpatched CPython:

```python
    import gc

    def test_task_del_collect(self):
        async def coro():
            return Evil()

        class Evil:
            def __del__(self):
                gc.collect()

        for i in range(100):
            task = self.new_task(self.loop, coro())
            self.loop.run_until_complete(task)
```

**Comment by 1st1** [comment]
The test also fails if we fix only Task, but not Future.

**Comment by thehesiod** [comment]
so I had something similar but it doesn't crash:
```python
import gc
import asyncio

class Evil:
    def __del__(self):
        gc.collect()


async def run():
    return Evil()


loop = asyncio.get_event_loop()

for i in range(100):
    loop.run_until_complete(asyncio.Task(run()))
#    f = asyncio.Future()
#    f.set_result(Evil())
```

**Comment by thehesiod** [comment]
ah, figured it out, had to have multiple tasks pending in the run loop.

File: Lib/test/test_asyncio/test_tasks.py
Comment (1st1) [inline Lib/test/test_asyncio/test_tasks.py]: This test won't run. `yield from` makes `test_task_del_collect()` a generator and test suite doesn't know how to run them.

Please change to 

```python
self.loop.run_until_complete(
    asyncio.gather(*[asyncio.Task(run()) for _ in range(100)]))
```

**Comment by thehesiod** [comment]
ya realized as I was going to work, this should fix it

**Review by 1st1 — APPROVED** [review]
I have no further comments, LGTM. @methane do you approve this PR?

File: Lib/test/test_asyncio/test_tasks.py
Comment (1st1) [inline Lib/test/test_asyncio/test_tasks.py]: Another nit: 'gc' import should be before 'functools' -- we sort imports alphabetically in CPython.

Comment (thehesiod) [reply Lib/test/test_asyncio/test_tasks.py]: may be worth running a linter on CI to catch these

Comment (1st1) [reply Lib/test/test_asyncio/test_tasks.py]: I agree, good idea. BTW help is always welcome by python/core-workflow. You're an awesome contributor, we'd love you to continue! :)

Comment (thehesiod) [reply Lib/test/test_asyncio/test_tasks.py]: you're too kind :)  Will keep that in mind if we run into any other cpython issues.

**Comment by thehesiod** [comment]
this set of PRs is a massive win for Python.  It would be nice if a core dev could look and find a way to nip it at the bud to avoid these hacky calls, as shown they're highly error prone.  Who knows how many third party modules are missing them too.

**Review by methane — APPROVED** [review]
LGTM

**Comment by 1st1** [comment]
Do we have C Task/Future in 3.5?

**Comment by methane** [comment]
No.  asyncio speedup is introduced at 3.6

**Pull request merged**
